L65-python-package-name.md

```diff
@@ -0,0 +1,132 @@
+Additional PyPI Packages for gRPC Python
+----
+* Author(s): lidiz
+* Approver: gnossen
+* Status: Draft
+* Implemented in: python
+* Last updated: 05/13/2020
+* Discussion at: **TBD**
+
+## Abstract
+
+gRPC Python is uploaded as `grpcio` on PyPI, but there is another package named
+`grpc`. Hence some users are confused. This document proposes gRPC Python to
+upload additional packages named `grpc` and `grpc-*` as proxies to our official
+packages.
+
+## Background
+
+### Package `grpc`
+
+Package `grpc` is an alternative version of gRPC Python only supports Python 2,
+and the first version is uploaded in 2013. We have seen complaints about the
+installation failure or unable to execute our example code. The tricky package
+name is a usability issue for years.
+
+### Package `grpcio`
+
+Package `grpcio` is first uploaded in 2015, given the fact that the name `grpc`
+already being taken, this name is likely a workaround.
+
+## Proposal
+
+### List of Package Name
+
+Here is a list of new proxy packages:
+* `grpc`
+* `grpc-status`
+* `grpc-channelz`
+* `grpc-tools`
+* `grpc-reflection`
+* `grpc-testing`
+* `grpc-health-checking`
+
+### Proxy Mechanism
+
+Each proxy package will raise a `RuntimeError` exception upon installation,
+asking users to install with the correct package name. For example, for proxy
+package `grpc`, the exception will look like:
+
+```
+RuntimeError: Please install the official package with: pip install grpcio
+```
+
+### Maintenance Cost
+
+The maintenance cost for this solution is minimum since these packages are
+agnostic to our releases.
+
+### Package Version
+
+All new proxy packages will be versioned as `1.0`.
```

Comment:
I didn't fully appreciate what this meant the first time I read through. I would reword to something more along the lines of "Rather than releasing new versions of the `grpc-*` packages in lockstep with the canonical `grpcio-*` packages for each release, we will instead upload only one new artifact for each package name, versioned as '1.0'.". You might also want to make an argument for this strategy. If we go with this strategy as opposed to uploading a new copy of the same `RunTimeError`-producing package, then people running `pip install grpc==1.29.0` will get a "package not found" error instead of a more informative "I think you have the wrong number"-style message.

Reply:
Thanks for the suggestion. Updated to:
```
Rather than releasing new versions of the `grpc` and `grpc-*` packages in
lockstep with the canonical `grpc` and `grpcio-*` packages for each release, we
will instead upload only one new artifact for each package name, versioned as
`v1.0`. A stable version number suggests this package's behavior won't change
actively.
```

I reword this section, so it is less confusing. For people who want to pin package version, they are using gRPC Python in a more serious scenario. I think they would need to double check the package name.

I'm also okay to upload all future versions with a script. But they would taken our ability to change its logic forever. Instrumenting our infrastructure to produce more artifacts for each release is time-costly, we might not want to make existing release procedure more complex.

Also, separated version suggests the nature of release agnostic for the new packages.

Reply:
> lockstep with the canonical `grpc` and `grpcio-*`

The current wording implies that both `grpc` and `grpcio` prefixes are canonical. How about `grpcio*`? That includes `grpcio`, `grpcio-channelz`, etc.

> I'm also okay to upload all future versions with a script. But they would taken our ability to change its logic forever

I don't think that's true. We can always decide that the next release of `grpc` will contain something other than a `RunTimeError`.

> Instrumenting our infrastructure to produce more artifacts for each release is time-costly, we might not want to make existing release procedure more complex.

Totally agree.